tests/common.py

```diff
@@ -340,14 +334,39 @@ def compare_forces(
     def compare_forces_gpu_vs_reference(
         self,
         x: NDArray,
-        params: NDArray,
+        params_arrays: Iterable[NDArray],
         box: NDArray,
-        lambdas: List[float],
         potential: generic.Potential,
         rtol: float,
         precision,
         atol: float = 1e-8,
     ):
         return self.compare_forces(
-            x, params, box, lambdas, potential.to_reference(), potential.to_gpu(), rtol, precision, atol
+            x, params_arrays, box, potential.to_reference(), potential.to_gpu(), rtol, precision, atol
         )
+
+
+def gen_nonbonded_params_with_4d_offsets(rng: np.random.Generator, params, w_min: float, w_max: float):
+    num_atoms, _ = params.shape
+
+    def params_with_w_coords(w_coords):
+        return jnp.asarray(params).at[:, 3].set(w_coords)
+
+    # all zero
+    yield params_with_w_coords(0.0)
+
+    # all w_max
+    yield params_with_w_coords(w_max)
+
+    # half zero, half w_max
+    w_coords = jnp.zeros(num_atoms)
+    w_coords = w_coords.at[-num_atoms // 2 :].set(w_max)
+    yield params_with_w_coords(w_coords)
+
+    # discrete random uniform in {0, w_max}
+    w_coords = rng.choice([0.0, w_max], (num_atoms,))
```

Comment:
Nit / curiosity: Motivation for set `{0, w_max}`?

(From context, expected a slightly different set `{0, w_min, w_max}`, or possibly redundant `{0, +w_min, +w_max, -w_min, -w_max}`.)

Reply:
The idea was just to test with sparse w coordinates (nonvanishing probability of zero), in contrast with the continuous uniform case (where there won't be any w that are exactly zero) and the ordered "half 0, half w_max" case. Since I was focused on "sparse, randomly allocated", I didn't think to test with other nonzero w values, but I don't see any harm in adding this.

Reply:
> Another way to accomplish this

Thanks, I think your suggestion is better. Updated in bc14618.